### DISCUSSION TITLE
Passing parameters to state machine in PIO

### DISCUSSION THREAD

**Discussion by Prakhar005** [micropython/micropython#9362] [Using MicroPython] [answered]
Hello all,

I am working on defining a state machine using the PIO functionality of raspberry Pico. I am using micropython for defining the program. In my state machine, I would like to set the scratch registers 'x' and 'y' by passing user-defined parameters to the state machine program. However, it seems there is no direct way to pass arguments to the state machine program. I was wondering if anyone has any hints on how to solve this issue? thank you!

**Comment by Vendelator** [comment, marked as answer]
If I understand what you want to do, you need to first put data into PIO and then PULL from inside an SM.

```py
import rp2

@rp2.asm_pio()
def pio_function():
    pull()
    mov(x, osr)

sm_0 = rp2.StateMachine(0, pio_function, freq=10_000)
user_parameter = int(input("Write user input: "))
sm_0.put(user_parameter)
```

Something along those lines.

BR

**Reply by Prakhar005** [reply]
Thank you for the suggested implementation. This makes a lot of sense! I use 'x' register as a counter which decrements to zero in the state machine. However, when the state machine wraps back to the start, I would like 'x' to be initialized again with the same user-specified value. I am unable to figure out how to re-initialize it. Is there a way to implement the 'put' command in the state machine itself and pass 'user_parameter' as an argument to 'pro_function()'. Thank you for the help! :)

**Reply by Vendelator** [reply]
I am guessing now. Because I am very much a beginner with this.

But if you:
```py
mov(y, osr)
mov(x, y)
```

y should not get "used up" and keep refilling x.


I saw in another example they did like this to keep refilling the program. I assume X will then change when a new value is put into it. (copied from OSR)
```py
@rp2.asm_pio(set_init=rp2.PIO.OUT_LOW)
def stepper():
    pull(noblock)
    mov(x, osr
    mov(y, x)

    jmp(not_x, "end")

    label("delay")
    jmp(x_dec, "delay")

    label("end")
    mov(x, y)
```

**Reply by Prakhar005** [reply]
Thank you for the help! I tried to implement your idea and it works very well. Thank you!

**Reply by Vendelator** [reply]
This I great news, I have learned from this as well.

Good luck with your coding.
